v2/client/reservation.go

```diff
@@ -0,0 +1,95 @@
+package client
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	pbv2 "github.com/libp2p/go-libp2p-circuit/v2/pb"
+	"github.com/libp2p/go-libp2p-circuit/v2/util"
+
+	"github.com/libp2p/go-libp2p-core/host"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/libp2p/go-libp2p-core/peerstore"
+)
+
+// Reservation is a struct carrying information about a relay/v2 slot reservation.
+type Reservation struct {
+	// Expiration is the expiration time of the reservation
+	Expiration time.Time
+	// Relay is the public addresses of the relay, which can be used for constructing
+	// and advertising relay specific addresses.
+	Relay peer.AddrInfo
+
+	// LimitDuration is the time limit for which the relay will keep a relayed connection
+	// open. If 0, there is no limit.
+	LimitDuration time.Duration
+	// LimitData is the number of bytes that the relay will relay in each direction before
+	// resetting a relayed connection.
+	LimitData int64
+
+	// TODO reservation voucher
+}
+
+// Reserve reserves a slot in a relay and returns the reservation information.
+// Clients must reserve slots in order for the relay to relay connections to them.
+func Reserve(ctx context.Context, h host.Host, ai peer.AddrInfo) (*Reservation, error) {
```

Comment:
It will make use of the Expiration/TTL so that it knows when to refresh.
The others may or may not be used, so why not include them? For instance, if those are missing then it's an unlimited relay, which is useful to know.

Reply:
Make sense.